### DISCUSSION TITLE
Differences with UXTerm wrt vim

### DISCUSSION THREAD

**Discussion by chdiza** [kovidgoyal/kitty#4204] [Q&A] [answered]
I'm switching from macOS to linux, and so I'm on an Apple laptop that is running Ubuntu 20.04.

I notice that when I use `uxterm` as my terminal, I have no trouble mapping the Option key inside of Vim.  "Option" key is its Apple name; it's the key just to the left of the left-side "Cmd" key.  Inside Vim, I can for example do `:nmap <M-t> :Myfavoritething<CR>` and it works just fine---when I use `uxterm`.

In kitty, those mappings don't work in Vim.   I don't know enough about terminals to tell whether the problem is with kitty or with Vim or with my Vim configuration.  (I note that the mappings fail even when Vim is started thus: `vim -u NONE`.)  Is there a kitty setting I'm unaware of?  Perhaps the kitty included in this version of Ubuntu (0.15.0) is too old and this has already been fixed.  Or is there a Vim setting I'm unaware of?

In `uxterm` I can even get fancy and have things like `:nmap <S-C-l>` and `:nmap <S-M-l>` work---and also be distinguished from `<C-l> and `<M-l>`.  Those mappings also work in gVim.

**Comment by Luflosi** [comment]
Execute `kitty --debug-keyboard` to see any key events kitty sees.

**Comment by chdiza** [comment]
Thanks, but I don't know how to read the mass of stuff that appears.  It *looks* like kitty thinks I hit `Alt_L` and then `t`.  But then Vim should've triggered the `nmap <M-t> :tabnew<CR>` that's in effect.  Instead, no tab is opened, and I have to hit ":" twice for some reason to get to the Vim cmdline.

```
Press xkb_keycode: 0x40 clean_sym: Alt_L composed_sym: Alt_L mods: none glfw_key: 342 (LEFT ALT) xkb_key: 65513 (Alt_L)
on_key_input: glfw key: 342 native_code: 0xffe9 action: PRESS mods: 0x0 text: '' state: 0 sent key to child
Press xkb_keycode: 0x1c clean_sym: t composed_sym: t mods: alt glfw_key: 84 (T) xkb_key: 116 (t)
on_key_input: glfw key: 84 native_code: 0x74 action: PRESS mods: 0x4 text: '' state: 0 sent key to child
Release xkb_keycode: 0x1c clean_sym: t mods: alt glfw_key: 84 (T) xkb_key: 116 (t)
on_key_input: glfw key: 84 native_code: 0x74 action: RELEASE mods: 0x4 text: '' state: 0 ignoring as keyboard mode does not allow release events
Release xkb_keycode: 0x40 clean_sym: Alt_L mods: alt glfw_key: 342 (LEFT ALT) xkb_key: 65513 (Alt_L)
on_key_input: glfw key: 342 native_code: 0xffe9 action: RELEASE mods: 0x4 text: '' state: 0 ignoring as keyboard mode does not allow release events
```

**Comment by kovidgoyal** [comment]
Running with up to date kitty and up to date vim alt-t works fine. Pressing ctrl+v in vim and then alt+t shows ^[t as expected. 0.15.0 is so old I have no clue what the status on that is. And as for extended key support see https://sw.kovidgoyal.net/kitty/keyboard-protocol/ kitty has the best key support of any terminal out there. vim doesnt (yet) support that protocol properly, but if you want an editor that does, try kakoune.

**Comment by chdiza** [comment]
> Pressing ctrl+v in vim and then alt+t shows ^[t as expected

Hmm, it shows that here on 0.15.0 as well.  And yet in `uxterm` Vim shows something else: a lowercase "o" but with a circumflex on it (0x00f4).  FYI this is Vim 8.2.3582.

Let me say that I find kitty a very promising candidate as a replacement for iTerm2!  Konsole and others have the same problem I'm reporting here, but kitty is *sooooo* much better than the others in pretty much every respect.

**Comment by kovidgoyal** [comment]
That would imply you are using a keyboard layout where alt+t is used to
generate some unicode text. For that you will need to update kitty.

**Comment by chdiza** [comment]
I tried the latest kitty release, to no avail.

> That would imply you are using a keyboard layout where alt+t is used to generate` some unicode text.

I *think* I'm just using whatever the Ubuntu installer chose.  "Apple Macbook/Macbook Pro" I think is what it's named."  I'm a bit confused as to why this Vim mapping works in `uxterm` if my keyboard layout is wonky.

**Comment by kovidgoyal** [comment]
On Mon, Nov 08, 2021 at 03:55:24AM -0800, chdiza wrote:
> I tried the latest kitty release, to no avail.
> 
> > That would imply you are using a keyboard layout where alt+t is used to generate` some unicode text.
> 
> I *think* I'm just using whatever the Ubuntu installer chose.  "Apple Macbook/Macbook Pro" I think is what it's named."  I'm a bit confused as to why this Vim mapping works in `uxterm` if my keyboard layout is wonky.

You want to see what bytes are being sent for a key use 

showkey -a

there is no way alt+t would generate a unicode character in a
default/pc101 layout.

**Comment by chdiza** [comment]
OK, thanks @kovidgoyal, I did `showkey -a` and Option+t.

In kitty this gives;
```
^[t 	27 0033 0x1b
 	116 0164 0x74
```
That's also what I get in konsole and sakura.  In uxterm it gives:
```
ô       195 0303 0xc3
        180 0264 0xb4
```
All the above is with my MacBookPro running Ubuntu 2.04 and its keyboard laout set (in KDE's systemsettings) as either of "Generic | Generic 101-key PC" or "Apple | MacBook/MacBook Pro".  They both produce the same.

Perhaps we should forget uxterm since what I really care about is why I can't map <M-t> in Vim inside kitty, and why (in Vim in kitty) that keychord produces an odd state wherein I have to hit ":" twice to get out of it.  Since this problem also happens to me in sakura and konsole too, I'm starting to suspect that there's some Vim option I should be setting.

**Comment by chdiza** [comment]
BTW we can add `urxvt` to the list of those that behave *un*like uxterm, both wrt `showkey -a` and attempting to map <M-t> in Vim.

**Comment by trygveaa** [comment]
For `<A-key>`/`<M-key>` mappings, vim assumes that pressing the Alt key sets the 8th bit of a typed character. xterm does this by default. See `:h :map-alt-keys` in vim for more info about this. For terminals where alt sends escape instead, like kitty, you can map it by using `<Esc>key` instead.

**Comment by chdiza** [comment]
> See :h :map-alt-keys in vim for more info about this.

Thanks.  I'd read that before and it's very discouraging (unless I go back to uxterm).  The gist of the fourth paragraph is that terminals that send the ESC prefix are out of luck.

> you can map it by using <Esc>key instead.

Thanks for that too!  I'm afraid that's going to be too easy to accidentally collide with.  (And it won't work in insert mode AFAIK.)

**Reply by trygveaa** [reply]
> Thanks for that too! I'm afraid that's going to be too easy to accidentally collide with.

What do you mean by easy to collide? That you accidentally press escape and then the rest of the mapping when you don't want to trigger that mapping?

>  (And it won't work in insert mode AFAIK.)

It works fine in insert mode. Having such a mapping will introduce a delay when pressing escape to go back to normal mode, which is a bit annoying, but you can press normal mappings immediately before the delay is over, so it doesn't matter that much.

**Comment by kovidgoyal** [comment, marked as answer]
On Tue, Nov 09, 2021 at 01:07:27PM -0800, chdiza wrote:
> > See :h :map-alt-keys in vim for more info about this.
> 
> Thanks.  I'd read that before and it's very discouraging (unless I go back to uxterm).  The gist of the fourth paragraph is that terminals that send the ESC prefix are out of luck.

Convince vim to enter the 21st century and implement support for
comprehensive keyboard handling:
https://sw.kovidgoyal.net/kitty/keyboard-protocol/

**Comment by chdiza** [comment]
> Convince vim to enter the 21st century and implement support for comprehensive keyboard handling:

I would love that.

(I hope nobody thought I have settled on blaming *kitty* for my difficulties.  I quickly realized that a whole bunch of other terminals behave like kitty in this respect.)

 To the extent that I understand what the heck is going on, the problem is that keyboard modifiers in terminals are a dumpster fire, and this gets baked into important programs, which are unable (unwilling?) to handle noble efforts at extinguishing the flames (like @kovidgoyal's).

**Reply by kovidgoyal** [reply]
Yeah essentially, the kitty keyboard protocol puts key handling in terminals to rest once and for all. Hopefully someday more software will be updated to support it. For the moment, one has to still suffer.
